str/associatr/multiple_testing_correction/run_gene_level_pval.py

```diff
@@ -0,0 +1,306 @@
+        raise ValueError('All p-values must be between 0 and 1!')
+
+    # check if there are p-values that are either exactly 0 or 1.
+    is_zero = (pvals == 0).sum() >= 1
+    is_one = (pvals == 1).sum() >= 1
+
+    # check the validity of weights (default: equal weights) and standardize them.
+    if weights is None:
+        weights = np.repeat(1 / len(pvals), len(pvals))
+    elif len(weights) != len(pvals):
+        raise ValueError(
+            'The length of weights should be the same as that of the p-values!'
+        )
+    elif (weights < 0).sum() > 0:
+        raise ValueError('All the weights must be positive!')
+    else:
+        weights = weights / np.sum(weights)
+
+    # check if there are very small non-zero p-values
+    is_small = pvals < 1e-16
+    if is_small.sum() == 0:
+        cct_stat = np.sum(weights * np.tan((0.5 - pvals) * np.pi))
+    else:
+        cct_stat = np.sum((weights[is_small] / pvals[is_small]) / np.pi)
+        cct_stat += np.sum(
+            weights[~is_small] * np.tan((0.5 - pvals[~is_small]) * np.pi)
+        )
+
+    if is_zero:
+        pval = 0
+    elif is_one:
+        print('There are p-values exactly equal to 1!')
+        pval = min(1, min(pvals) * len(pvals))
+
+    # check if the test statistic is very large.
+    elif cct_stat > 1e15:
+        pval = (1 / cct_stat) / np.pi
+    else:
+        pval = 1 - cauchy.cdf(cct_stat)
+
+    # write to output
+    gcs_output = output_path(
+        f'gene_level_pvals/acat/{cell_type}/chr{chromosome}/{gene_name}_gene_level_pval.tsv',
+        'analysis',
+    )
+    with to_path(gcs_output).open('w') as f:
+        f.write(
+            f'gene_name\tgene_level_pval\tchr\tpos\tn_samples_tested\tlowest_raw_pval\tcoeff\tse\tr2\tmotif\tref_len\tallele_freq\n'
+        )
+        f.write(
+            f'{gene_name}\t{str(pval)}\t{str(chr)}\t{str(pos)}\t{str(n_samples_tested)}\t{str(raw_pval)}\t{str(coeff)}\t{str(se)}\t{str(r2)}\t{str(motif)}\t{str(ref_len)}\t{str(allele_frequency)}\n'
+        )
+
+
+def bonferroni_compute(
+    gene_name,
+    pvals,
+    cell_type,
+    chromosome,
+    chr,
+    pos,
+    n_samples_tested,
+    raw_pval,
+    coeff,
+    se,
+    r2,
+    motif,
+    ref_len,
+    allele_frequency,
+):
+    """
+    Computes Bonferroni adjusted p-value of the lowest raw p-value for a gene
+    """
+    pval = min(pvals) * len(pvals)
+    # write to output
+    gcs_output = output_path(
+        f'gene_level_pvals/bonferroni/{cell_type}/chr{chromosome}/{gene_name}_gene_level_pval.tsv',
+        'analysis',
+    )
+    with to_path(gcs_output).open('w') as f:
+        f.write(
+            f'gene_name\tgene_level_pval\tchr\tpos\tn_samples_tested\tlowest_raw_pval\tcoeff\tse\tr2\tmotif\tref_len\tallele_freq\n'
+        )
+        f.write(
+            f'{gene_name}\t{str(pval)}\t{str(chr)}\t{str(pos)}\t{str(n_samples_tested)}\t{str(raw_pval)}\t{str(coeff)}\t{str(se)}\t{str(r2)}\t{str(motif)}\t{str(ref_len)}\t{str(allele_frequency)}\n'
```

Comment:
you don't need to wrap these elements in string translations - using an f-string will use either the object's `repr` or `str` methods to get a string representation.